File: eBUS_SDK/eBUS_4_1_5_3702/SamplesDotNet/SimpleApplication/MainForm.cs
// *****************************************************************************
//
//     Copyright (c) 2013, Pleora Technologies Inc., All rights reserved.
//
// *****************************************************************************

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.Diagnostics;
using PvDotNet;
using PvGUIDotNet;


namespace SimpleApplication
{
    public partial class MainForm : Form
    {
        /// <summary>
        /// Constructor
        /// </summary>
        public MainForm()
        {
            InitializeComponent();

            // Handlers used to callbacks in the UI thread
            mDisconnectedHandler += new GenericHandler(OnDisconnected);
            mAcquisitionModeChangedHandler += new GenericHandler(OnAcquisitionModeChanged);
            mAcquisitionStateChangedHandler += new GenericHandler(OnAcquisitionStateChanged);

            // Set browser form owners
            mCommBrowser.Owner = this;
            mDeviceBrowser.Owner = this;
            mStreamBrowser.Owner = this;

            // Create display thread, hook display event
            mDisplayThread = new PvDisplayThread();
            mDisplayThread.OnBufferDisplay += new OnBufferDisplay(OnBufferDisplay);
        }

        // Handler used to bring link disconnected event in the main UI thread
        private delegate void GenericHandler();
        private GenericHandler mDisconnectedHandler = null;

        // Handler used to bring acquisition mode changed event in the main UI thread
        private GenericHandler mAcquisitionModeChangedHandler = null;

        // Handler used to bring acquisition state manager callbacks in the main UI thread
        private GenericHandler mAcquisitionStateChangedHandler = null;

        // Main application objects: device, stream, pipeline
        private PvDevice mDevice = null;
        private PvStream mStream = null;
        private PvPipeline mPipeline = null;

        // Acquisition state manager
        private PvAcquisitionStateManager mAcquisitionManager = null;

        // Display thread
        private PvDisplayThread mDisplayThread = null;

        // Parameter browsers
        private BrowserForm mCommBrowser = new BrowserForm();
        private BrowserForm mDeviceBrowser = new BrowserForm();
        private BrowserForm mStreamBrowser = new BrowserForm();

        /// <summary>
        /// Method that syncs the whole UI enabled state.
        /// </summary>
        private void EnableInterface()
        {
            // Connect and disconnect button
            connectButton.Enabled = (mDevice == null);
            disconnectButton.Enabled = (mDevice != null);

            // Tree browser buttons
            EnableTreeBrowsers();

            // Acquisition controls
            EnableControls();
        }

        /// <summary>
        /// Connect button click handler.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void connectButton_Click(object sender, EventArgs e)
        {
            // Create a device finder wnd and open the select device dialog
            PvDeviceFinderForm lFinder = new PvDeviceFinderForm();

            // Show device finder
            if ((lFinder.ShowDialog() != DialogResult.OK) ||
                (lFinder.Selected == null))
            {
                return;
            }

            // Display hourglass
            Cursor lOld = Cursor;
            Cursor = Cursors.WaitCursor;

            // Connect to device
            Connect(lFinder.Selected);

            // Restore cursor
            Cursor = lOld;
        }

        /// <summary>
        /// Disconnect button event handler.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void disconnectButton_Click(object sender, EventArgs e)
        {
            // Show hourglass
            Cursor lOld = Cursor;
            Cursor = Cursors.WaitCursor;

            // Disconnect from device
            Disconnect();

            // Restore cursor
            Cursor = lOld;
        }

        /// <summary>
        /// Enable the tree browser buttons.
        /// </summary>
        private void EnableTreeBrowsers()
        {
            // Set the state of the Communication, Device and Image Stream control button
            communicationButton.Enabled = (mDevice != null);
            deviceButton.Enabled = (mDevice != null);
            streamButton.Enabled = (mDevice != null);
        }

        /// <summary>
        /// Enables the acquisition controls
        /// </summary>
        private void EnableControls()
        {
            if (mAcquisitionManager == null)
            {
                // Not connected: acquisition state manager not available, directly disable all
                modeComboBox.Enabled = false;
                playButton.Enabled = false;
                stopButton.Enabled = false;
            }
            else
            {
                // Read current acquisition state.
                bool lLocked = mAcquisitionManager.State == PvAcquisitionState.Unlocked;

                // Mode and play are enabled if not locked
                modeComboBox.Enabled = lLocked;
                playButton.Enabled = lLocked;

                // Stop is enabled only if locked
                stopButton.Enabled = !lLocked;
            }
        }

        /// <summary>
        /// Form load event handler. 
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void MainForm_Load(object sender, EventArgs e)
        {
            // Just syncs the enabled state of the UI
            EnableInterface();
        }

        /// <summary>
        /// Form closing handler
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void MainForm_FormClosing(object sender, FormClosingEventArgs e)
        {
            Disconnect();
        }

        /// <summary>
        /// Connects to the device from a device info
        /// </summary>
        /// <param name="aDI"></param>
        void Connect(PvDeviceInfo aDI)
        {
            // Just in case we came here still connected...
            Disconnect();
           
            if (aDI == null)
            {
                MessageBox.Show("GigE Vision devices not currently supported by this sample.", Text);
                return;
            }            

            // Create and connect the device controller based on the selected device
            mDevice = PvDevice.CreateAndConnect(aDI);

            try
            {
                // Create and open stream
                mStream = PvStream.CreateAndOpen(aDI.ConnectionID);

                // GigE Vision specific connection steps
                if (aDI.Type == PvDeviceInfoType.GEV)
                {
                    PvDeviceGEV lDeviceGEV = mDevice as PvDeviceGEV;
                    PvStreamGEV lStreamGEV = mStream as PvStreamGEV;

                    // Negotiate packet size
                    lDeviceGEV.NegotiatePacketSize();

                    // Set stream destination to our stream object
                    lDeviceGEV.SetStreamDestination(lStreamGEV.LocalIPAddress, lStreamGEV.LocalPort);
                }

                // Create pipeline - requires stream
                mPipeline = new PvPipeline(mStream);
            }
            catch (PvException ex)
            {
                // Failure at some point, display and abort
                MessageBox.Show(ex.Message, Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
                Disconnect();

                return;
            }

            // Register to all events of the parameters in the device's node map
            foreach (PvGenParameter lParameter in mDevice.Parameters)
            {
                lParameter.OnParameterUpdate += new OnParameterUpdateHandler(OnParameterChanged);
            }

            // Connect link disconnection handler
            mDevice.OnLinkDisconnected += new OnLinkDisconnectedHandler(OnLinkDisconnected);

            // Update device attributes
            UpdateAttributes(aDI);

            // Fill acquisition mode combo box
            modeComboBox.Items.Clear();
            PvGenEnum lMode = mDevice.Parameters.GetEnum("AcquisitionMode");
            if (lMode != null)
            {
                foreach (PvGenEnumEntry lEntry in lMode)
                {
                    if (lEntry.IsAvailable)
                    {
                        int lIndex = modeComboBox.Items.Add(lEntry.ValueString);
                        if (lEntry.ValueInt == lMode.ValueInt)
                        {
                            modeComboBox.SelectedIndex = lIndex;
                        }
                    }
                }
            }

            // Ready image reception
            StartStreaming();

            // Sync the UI with our new status
            EnableInterface();
        }

        /// <summary>
        /// Updates the device attributes on the form
        /// </summary>
        /// <param name="aDI"></param>
        void UpdateAttributes(PvDeviceInfo aDI)
        {
            string lVendorName = "";
            string lModelName = "";
            string lMACAddress = "";
            string lIPAddress = "";
            string lGUID = "";
            string lUserDefinedName = "";

            if (aDI != null)
            {
                // Get device attributes
                lVendorName = aDI.VendorName;
                lModelName = aDI.ModelName;
                lMACAddress = "N/A";
                lIPAddress = "N/A";
                lGUID = "N/A";
                lUserDefinedName = aDI.UserDefinedName;

                // GigE Vision specific device attributes
                PvDeviceInfoGEV lDeviceInfoGEV = aDI as PvDeviceInfoGEV;
                if (lDeviceInfoGEV != null)
                {
                    lMACAddress = lDeviceInfoGEV.MACAddress;
                    lIPAddress = lDeviceInfoGEV.IPAddress;
                }

                // USB3 Vision specific device attributes
                PvDeviceInfoU3V lDeviceInfoU3V = aDI as PvDeviceInfoU3V;
                if (lDeviceInfoU3V != null)
                {
                    lGUID = lDeviceInfoU3V.DeviceGUID;
                }
            }

            // Fill device attribute text boxes
            manufacturerTextBox.Text = lVendorName;
            modelTextBox.Text = lModelName;
            macAddressTextBox.Text = lMACAddress;
            ipAddressTextBox.Text = lIPAddress;
            guidTextBox.Text = lGUID;
            nameTextBox.Text = lUserDefinedName;
        }

        /// <summary>
        /// Disconnects from the device
        /// </summary>
        private void Disconnect()
        {
            // Close all configuration child windows
            mDeviceBrowser.Browser.GenParameterArray = null;
            CloseGenWindow(mDeviceBrowser);
            mStreamBrowser.Browser.GenParameterArray = null;
            CloseGenWindow(mStreamBrowser);

            if (mStream != null)
            {
                // If streaming, stop streaming
                if (mStream.IsOpen)
                {
                    StopStreaming();
                    
                    mStream.Close();
                    mStream = null;
                }
            }

            if (mDevice != null)
            {
                if (mDevice.IsConnected)
                {
                    // Disconnect events.
                    mDevice.OnLinkDisconnected -= new OnLinkDisconnectedHandler(OnLinkDisconnected);
                    foreach (PvGenParameter lP in mDevice.Parameters)
                    {
                        lP.OnParameterUpdate -= new OnParameterUpdateHandler(OnParameterChanged);
                    }

                    mDevice.Disconnect();
                    mDevice = null;
                }
            }
            

            display.Clear();
            UpdateAttributes(null);
            statusControl.Warning = "";

            // Sync the UI with our new status
            EnableInterface();
        }

        /// <summary>
        /// GenICam parameter invalidation event, registered for all parameters.
        /// </summary>
        /// <param name="aParameter"></param>
        void OnParameterChanged(PvGenParameter aParameter)
        {
            string lName = aParameter.Name;
            if (lName == "AcquisitionMode")
            {
                // Have main UI thread update the acquisition mode combo box status
                BeginInvoke(mAcquisitionModeChangedHandler);
            }
        }

        /// <summary>
        /// Acquisition mode event handler in main thread.
        /// </summary>
        private void OnAcquisitionModeChanged()
        {
            // Get parameter
            PvGenEnum lParameter = mDevice.Parameters.GetEnum("AcquisitionMode");

            // Update value: find which matches in the combo box
            string lValue = lParameter.ValueString;
            foreach (string lString in modeComboBox.Items)
            {
                if (lValue == lString)
                {
                    modeComboBox.SelectedValue = lString;
                }
            }
        }

        /// <summary>
        /// Direct acquisition state changed handler. Bring back to main UI thread.
        /// </summary>
        /// <param name="aDevice"></param>
        /// <param name="aStream"></param>
        /// <param name="aSource"></param>
        /// <param name="aState"></param>
        void OnAcquisitionStateChanged(PvDevice aDevice, PvStream aStream, uint aSource, PvAcquisitionState aState)
        {
            // Invoke event in main UI thread.
            BeginInvoke(mAcquisitionStateChangedHandler);
        }
        
        /// <summary>
        /// OnAcquisitionStateChanged main UI thread counterpart. Syncs acquisition control status.
        /// </summary>
        private void OnAcquisitionStateChanged()
        {
            EnableControls();
        }

        /// <summary>
        /// Displays a GenICam browser form. If already visible, close it for toggle effect.
        /// </summary>
        /// <param name="aForm"></param>
        /// <param name="aParams"></param>
        /// <param name="aTitle"></param>
        private void ShowGenWindow(BrowserForm aForm, PvGenParameterArray aParams, string aTitle)
        {
            if (aForm.Visible)
            {
                CloseGenWindow(aForm);
                return;
            }

            // Creates, assigns parameters, set title and show modeless
            aForm.Text = aTitle;
            aForm.Browser.GenParameterArray = aParams;
            aForm.Show();
        }

        /// <summary>
        /// Closes a GenICam browser form.
        /// </summary>
        /// <param name="aForm"></param>
        private void CloseGenWindow(Form aForm)
        {
            aForm.Hide();
        }

        /// <summary>
        /// Direct device disconnect handler. Just jump back to main UI thread.
        /// </summary>
        /// <param name="aDevice"></param>
        private void OnLinkDisconnected(PvDevice aDevice)
        {
            BeginInvoke(mDisconnectedHandler);
        }

        /// <summary>
        /// Reaction to device disconnected event: stop streaming, close device connection.
        /// </summary>
        private void OnDisconnected()
        {
            MessageBox.Show("Connection to device lost.", Text, MessageBoxButtons.OK, MessageBoxIcon.Information);

            StopStreaming();
            Disconnect();
        }

        /// <summary>
        /// Setups streaming. After calling this method the application is ready to receive data.
        /// StartAcquisition will instruct the device to actually start sending data.
        /// </summary>
        private void StartStreaming()
        {
            // Configure status control
            statusControl.Stream = mStream;
            statusControl.DisplayThread = mDisplayThread;

            // Start threads
            mDisplayThread.Start(mPipeline, mDevice.Parameters);
            mDisplayThread.Priority = PvThreadPriority.AboveNormal;

            // Configure acquisition state manager
            mAcquisitionManager = new PvAcquisitionStateManager(mDevice, mStream);
            mAcquisitionManager.OnAcquisitionStateChanged += new OnAcquisitionStateChanged(OnAcquisitionStateChanged);

            // Start pipeline
            mPipeline.Start();
        }

        /// <summary>
        /// Callback from display thread delegating buffer display task.
        /// </summary>
        /// <param name="aDisplayThread">Display thread</param>
        /// <param name="aBuffer">Buffer to display</param>
        void OnBufferDisplay(PvDisplayThread aDisplayThread, PvBuffer aBuffer)
        {
            display.Display(aBuffer);
        }

        /// <summary>
        /// Stops streaming. After calling this method the application is no longer armed or ready
        /// to receive data.
        /// </summary>
        private void StopStreaming()
        {
            if (!mDisplayThread.IsRunning)
            {
                return;
            }

            // Stop display thread
            mDisplayThread.Stop(false);

            // Release acquisition manager
            mAcquisitionManager.Dispose();
            mAcquisitionManager = null;

            // Stop pipeline
            if (mPipeline.IsStarted)
            {
                mPipeline.Stop();
            }

            // Wait on display thread
            mDisplayThread.WaitComplete();
        }

        /// <summary>
        /// Starts acquisition.
        /// </summary>
        private void StartAcquisition()
        {
            // Get payload size
            UInt32 lPayloadSize = lPayloadSize = mDevice.PayloadSize;

            // Propagate to pipeline to make sure buffers are big enough
            mPipeline.BufferSize = lPayloadSize;

            // Reset pipeline
            mPipeline.Reset();

            // Reset stream statistics
            PvGenCommand lResetStats = mStream.Parameters.GetCommand("Reset");
            lResetStats.Execute();

            // Reset display thread stats (mostly frames displayed per seconds)
            mDisplayThread.ResetStatistics();

            // Use acquisition manager to send the acquisition start command to the device
            mAcquisitionManager.Start();
        }

        /// <summary>
        /// Stops the acquisition.
        /// </summary>
        private void StopAcquisition()
        {
            // Use acquisition manager to send the acquisition stop command to the device
            mAcquisitionManager.Stop();
        }

        /// <summary>
        /// Communication button event handler. Display communication parameters browser form.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void communicationButton_Click(object sender, EventArgs e)
        {
            ShowGenWindow(mCommBrowser, mDevice.CommunicationParameters, "Communication Control");
        }

        /// <summary>
        /// Device button event handler. Display device parameters browser form.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void deviceButton_Click(object sender, EventArgs e)
        {
            ShowGenWindow(mDeviceBrowser, mDevice.Parameters, "Device Control");
        }

        /// <summary>
        /// Stream button event handler. Display stream parameters browser form.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void streamButton_Click(object sender, EventArgs e)
        {
            ShowGenWindow(mStreamBrowser, mStream.Parameters, "Image Stream Control");
        }

        /// <summary>
        /// Acquisition mode combo box changed. Propagate to device.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void modeComboBox_SelectedIndexChanged(object sender, EventArgs e)
        {
            // Nothing selected? Unexpected, just do nothing...
            if (modeComboBox.SelectedIndex < 0)
            {
                return;
            }

            PvGenEnum lMode = mDevice.Parameters.GetEnum("AcquisitionMode");

            // Take selection, propagate to parameter
            string lSelected = (string)modeComboBox.SelectedItem;
            lMode.ValueString = lSelected;
        }

        /// <summary>
        /// Play button click handler. Start acquisition.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void playButton_Click(object sender, EventArgs e)
        {
            StartAcquisition();
        }

        /// <summary>
        /// Stop button click handler. Stop acquisition.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void stopButton_Click(object sender, EventArgs e)
        {
            StopAcquisition();
        }
    }
}

File: eBUS_SDK/eBUS_4_1_5_3702/SamplesDotNet/SimpleApplicationVB/MainForm.vb
' *****************************************************************************
'
'     Copyright (c) 2013, Pleora Technologies Inc., All rights reserved.
'
' *****************************************************************************

Public Class MainForm

    Public Sub New()
        InitializeComponent()

        ' Handlers used to callbacks in the UI thread.
        mDisconnectedHandler = New GenericHandler(AddressOf OnDisconnected)
        mAcquisitionModeChangedHandler = New GenericHandler(AddressOf OnAcquisitionModeChanged)
        mAcquisitionStateChangedHandler = New GenericHandler(AddressOf OnAcquisitionStateChanged)

        ' Set browser form owners
        mCommBrowser.Owner = Me
        mDeviceBrowser.Owner = Me
        mStreamBrowser.Owner = Me

        ' Create display thread
        mDisplayThread = New PvDisplayThread()
    End Sub

    ' Declare delegate GenericHandler.
    Delegate Sub GenericHandler()

    ' Handler used to bring link disconnected event in the main UI thread.
    Dim mDisconnectedHandler As GenericHandler

    ' Handler used to bring acquisition mode changed event in the main UI thread.
    Dim mAcquisitionModeChangedHandler As GenericHandler

    ' Handler used to bring acquisition state manager callbacks in the main UI thread.
    Dim mAcquisitionStateChangedHandler As GenericHandler

    ' Initializes objects for PvDevice, PvStream, PvPipeline.
    Private mDevice As PvDevice = Nothing
    Private mStream As PvStream = Nothing
    Private mPipeline As PvPipeline = Nothing

    ' Acquisition state manager.
    Private mAcquisitionManager As PvAcquisitionStateManager = Nothing

    ' Display thread
    Private mDisplayThread As PvDisplayThread = Nothing

    ' Parameter browsers.
    Private mCommBrowser = New BrowserForm
    Private mDeviceBrowser = New BrowserForm
    Private mStreamBrowser = New BrowserForm

    ' Subroutine that set the state of the GUI.
    Private Sub SetGUIState()
        ' Set the state of connect and disconnect button.
        connectButton.Enabled = IsNothing(mDevice)
        disconnectButton.Enabled = Not IsNothing(mDevice)

        ' Set GigE Vision control settings.
        SetGenControlSettings()

        ' Set Acquisition controls.
        SetAcquisitionControl()
    End Sub

    ' Set the state of the Communication, Device and Image Stream control buttons.
    Private Sub SetGenControlSettings()
        ' Set the state of the communication, Device and Image Stream control button.
        communicationButton.Enabled = Not IsNothing(mDevice)
        deviceButton.Enabled = Not IsNothing(mDevice)
        streamButton.Enabled = Not IsNothing(mStream)
    End Sub

    ' Set the state of the acquisition control buttons.
    Private Sub SetAcquisitionControl()
        If Me.InvokeRequired Then
            Me.Invoke(New MethodInvoker(AddressOf SetAcquisitionControl))
        Else
            If IsNothing(mAcquisitionManager) Then
                ' Device is not connected so disables all acquisition controls.
                modeComboBox.Enabled = False
                playButton.Enabled = False
                stopButton.Enabled = False
            Else
                ' Read current acquisition state.
                Dim lLocked As Boolean = (mAcquisitionManager.State = PvAcquisitionState.Unlocked)

                ' Mode and play are enabled if not locked.
                modeComboBox.Enabled = lLocked
                playButton.Enabled = lLocked

                ' Stop is enabled only if locked.
                stopButton.Enabled = Not lLocked
            End If
        End If
    End Sub

    ' Connects to the device from a device info.
    Private Sub Connect(ByVal aDI As PvDeviceInfo)
        ' Just in case we came here still connected...
        Disconnect()

        Try

            ' Connect to device using device info.
            mDevice = PvDevice.CreateAndConnect(aDI)

            ' Open stream using device IP address.
            mStream = PvStream.CreateAndOpen(aDI.ConnectionID)

            ' Perform GigE Vision specific initialization
            If aDI.Type = PvDeviceInfoType.GEV Then

                Dim lDeviceGEV As PvDeviceGEV = TryCast(mDevice, PvDeviceGEV)
                Dim lStreamGEV As PvStreamGEV = TryCast(mStream, PvStreamGEV)

                ' Negotiate packet size
                lDeviceGEV.NegotiatePacketSize()

                ' Set stream destination to our stream object.
                lDeviceGEV.SetStreamDestination(lStreamGEV.LocalIPAddress, lStreamGEV.LocalPort)

            End If

            ' Create pipeline - requires stream.
            mPipeline = New PvPipeline(mStream)

        Catch lEx As PvException

            ' Display error message, disconnect device and abort.
            MessageBox.Show(lEx.Message, Text, MessageBoxButtons.OK, MessageBoxIcon.Error)
            Disconnect()
            Return

        End Try

        If mDevice.IsConnected Then

            Try

                ' Register to all events of the parameters in the device's node map.
                For Each lParameter As PvGenParameter In mDevice.Parameters
                    AddHandler lParameter.OnParameterUpdate, AddressOf OnParameterChanged
                Next

                ' Connect device disconnect notification.
                AddHandler mDevice.OnLinkDisconnected, AddressOf OnLinkDisconnected

                ' Update the UI attributes
                UpdateAttributes(aDI)

                ' Add items in acquisition mode combo box.
                modeComboBox.Items.Clear()
                Dim lMode As PvGenEnum = mDevice.Parameters.GetEnum("AcquisitionMode")
                If IsNothing(lMode) <> True Then

                    For Each lEntry As PvGenEnumEntry In lMode

                        If IsNothing(lEntry) <> True Then

                            Dim lIndex As Integer = modeComboBox.Items.Add(lEntry.ValueString)
                            If lEntry.ValueInt = lMode.ValueInt Then
                                modeComboBox.SelectedIndex = lIndex
                            End If

                        End If

                    Next

                End If

            Catch lEx As PvException

                MessageBox.Show(lEx.Message, Text, MessageBoxButtons.OK, MessageBoxIcon.Error)
                Disconnect()
                Return

            End Try

        End If

        If mStream.IsOpen Then
            ' Start receiving image.
            StartStreaming()
        End If

        ' Set the GUI according to our new status.
        SetGUIState()

    End Sub

    ' Update device attributes on the GUID
    Private Sub UpdateAttributes(ByVal aDI As PvDeviceInfo)

        Dim lVendorName As String = ""
        Dim lModelName As String = ""
        Dim lMACAddress As String = ""
        Dim lIPAddress As String = ""
        Dim lGUID As String = ""
        Dim lUserDefinedName As String = ""

        If Not IsNothing(aDI) Then

            ' Get device attributes
            lVendorName = aDI.VendorName
            lModelName = aDI.ModelName
            lMACAddress = "N/A"
            lIPAddress = "N/A"
            lGUID = "N/A"
            lUserDefinedName = aDI.UserDefinedName

            ' GigE Vision specific device attributes
            Dim lDeviceInfoGEV As PvDeviceInfoGEV = TryCast(aDI, PvDeviceInfoGEV)
            If Not IsNothing(lDeviceInfoGEV) Then

                lMACAddress = lDeviceInfoGEV.MACAddress
                lIPAddress = lDeviceInfoGEV.IPAddress

            End If

            ' USB3 Vision specific device attributes
            Dim lDeviceInfoU3V As PvDeviceInfoU3V = TryCast(aDI, PvDeviceInfoU3V)
            If Not IsNothing(lDeviceInfoU3V) Then

                lGUID = lDeviceInfoU3V.DeviceGUID

            End If

        End If

        ' Fill device attribute text boxes
        manufacturerTextBox.Text = lVendorName
        modelTextBox.Text = lModelName
        macAddressTextBox.Text = lMACAddress
        ipAddressTextBox.Text = lIPAddress
        guidTextBox.Text = lGUID
        nameTextBox.Text = lUserDefinedName

    End Sub

    ' Disconnects from the device.
    Private Sub Disconnect()

        ' Close all configuration child windows.
        mDeviceBrowser.Browser.GenParameterArray = Nothing
        CloseGenWindow(mDeviceBrowser)
        mStreamBrowser.Browser.GenParameterArray = Nothing
        CloseGenWindow(mStreamBrowser)
        If IsNothing(mStream) <> True Then
            ' If streaming, stop streaming.
            If mStream.IsOpen = True Then
                StopStreaming()
                mStream.Close()
                mStream = Nothing
            End If
        End If
        If IsNothing(mDevice) <> True Then
            If mDevice.IsConnected = True Then
                ' Disconnect events.
                RemoveHandler mDevice.OnLinkDisconnected, AddressOf OnLinkDisconnected
                For Each lP As PvGenParameter In mDevice.Parameters
                    RemoveHandler lP.OnParameterUpdate, AddressOf OnParameterChanged
                Next

                mDevice.Disconnect()
                mDevice = Nothing
            End If
        End If

        display.Clear()
        UpdateAttributes(Nothing)
        statusControl.Warning = ""

        ' Sync the UI with our new status.
        SetGUIState()
    End Sub

    ' Setups streaming. After calling this method the application is ready to receive data.
    ' StartAcquisition will instruct the device to actually start sending data.
    Private Sub StartStreaming()
        ' Configure display thread.
        AddHandler mDisplayThread.OnBufferDisplay, AddressOf OnBufferDisplay

        ' Configure status control.
        statusControl.Stream = mStream
        statusControl.DisplayThread = mDisplayThread

        ' Start threads.
        mDisplayThread.Start(mPipeline, mDevice.Parameters)
        mDisplayThread.Priority = PvThreadPriority.AboveNormal

        ' Configure acquisition state manager.
        mAcquisitionManager = New PvAcquisitionStateManager(mDevice, mStream)
        AddHandler mAcquisitionManager.OnAcquisitionStateChanged, AddressOf OnAcquisitionStateChanged

        ' Start pipeline.
        mPipeline.Start()
    End Sub

    ' Stops streaming. After calling this method the application is no longer armed or ready to receive data.
    Private Sub StopStreaming()
        ' Are we streaming?
        If Not mDisplayThread.IsRunning Then
            Return
        End If

        ' Stop display thread.
        mDisplayThread.Stop(False)

        ' Release acquisition manager.
        mAcquisitionManager.Dispose()
        mAcquisitionManager = Nothing
        If IsNothing(mPipeline) <> True Then
            ' Stop pipeline
            If mPipeline.IsStarted Then
                mPipeline.Stop()
            End If
        End If
        ' Wait on and release display thread.
        mDisplayThread.WaitComplete()
    End Sub

    ' Starts acquisition.
    Private Sub StartAcquisition()
        ' Get payload size
        Dim lPayloadSize = mDevice.PayloadSize

        ' Propagate to pipeline to make sure buffers are big enough.
        mPipeline.BufferSize = lPayloadSize

        ' Reset pipeline.
        mPipeline.Reset()

        ' Reset stream statistics.
        Dim lResetStats As PvGenCommand = mStream.Parameters.GetCommand("Reset")
        lResetStats.Execute()

        ' Reset display thread stats ( mostly frames displayed per seconds).
        mDisplayThread.ResetStatistics()

        ' Use acquisition manager to send the acquisition start command to the device.
        mAcquisitionManager.Start()
    End Sub

    ' Stops the acquisition.
    Private Sub StopAcquisition()
        Try
            ' Use acquisition manager to send the acquisition stop command to the device.
            mAcquisitionManager.Stop()
        Catch lEx As PvException
            MessageBox.Show(lEx.Message, Text, MessageBoxButtons.OK, MessageBoxIcon.Error)
        End Try

    End Sub

    ' Callback from display thread requesting a buffer to be displayed.
    Private Sub OnBufferDisplay(aDisplayThread As PvDisplayThread, aBuffer As PvBuffer)
        display.Display(aBuffer)
    End Sub

    ' Displays a GenICam browser form. If already visible, close it for toggle effect.
    Private Sub ShowGenWindow(ByVal aForm As BrowserForm, ByVal aParams As PvGenParameterArray, ByVal aTitle As String)
        If aForm.Visible Then
            CloseGenWindow(aForm)
            Return
        End If

        ' Create, assigned parameters, set title and show modeless.
        aForm.Text = aTitle
        aForm.Browser.GenParameterArray = aParams
        aForm.Show()
    End Sub

    ' Closes a GenICam browser form.
    Private Sub CloseGenWindow(ByVal aForm As Form)
        aForm.Hide()
    End Sub

    'Reaction to device disconnected event: stop streaming, close device connection.
    Private Sub OnDisconnected()
        MessageBox.Show("Connection to device lost.", Text)
        StopStreaming()
        Disconnect()
    End Sub

    ' Direct device disconnect handler. Just jump back to main UI thread.
    Private Sub OnLinkDisconnected(ByVal aDevice As PvDevice)
        BeginInvoke(mDisconnectedHandler)
    End Sub

    ' GenICam parameter invalidation event, registered for all parameters.
    Private Sub OnParameterChanged(ByVal aParameter As PvGenParameter)
        Dim lName As String = aParameter.Name
        If lName = "AcquisitionMode" Then
            ' Have main UI thread update the acquisition mode combo box status.
            BeginInvoke(mAcquisitionModeChangedHandler)
        End If
    End Sub

    ' Acquisition mode event handler in main thread.
    Private Sub OnAcquisitionModeChanged()
        ' Get parameter
        Dim lParameter As PvGenEnum = mDevice.Parameters.GetEnum("AcquisitionMode")

        ' Update value: find which matches in the combo box.
        Dim lValue As String = lParameter.ValueString
        For Each lString As String In modeComboBox.Items
            If lValue = lString Then
                modeComboBox.SelectedValue = modeComboBox.Items.ToString()
            End If
        Next
    End Sub

    ' Direct acquisition state changed handler. Bring back to main UI thread.
    Private Sub OnAcquisitionStateChanged(ByVal aDevice As PvDevice, ByVal aStream As PvStream, ByVal aSource As UInt32, ByVal aState As PvAcquisitionState)
        'Invoke event in main UI thread
        BeginInvoke(mAcquisitionStateChangedHandler)
    End Sub

    'OnAcquisitionStateChanged main UI thread counterpart. Syncs acquisition control status.
    Private Sub OnAcquisitionStateChanged()
        SetAcquisitionControl()
    End Sub

    Private Sub MainForm_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        ' Just syncs the enabled state of the UI.
        SetGUIState()
    End Sub

    Private Sub btnConnect_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles connectButton.Click
        ' create a device finder form and open the select device dialog.
        Dim lFinder As PvDeviceFinderForm = New PvDeviceFinderForm()
        Try
            ' Show device finder
            If ((lFinder.ShowDialog() <> DialogResult.OK) OrElse (IsNothing(lFinder.Selected) = True)) Then
                Return
            End If

            'Display hourglass.
            Dim lPrevCursor As Cursor = Cursor
            Cursor = Cursors.WaitCursor

            ' Connect to device.
            Connect(lFinder.Selected)

            ' Restore cursor.
            Cursor = lPrevCursor
        Catch lEx As Exception
            MessageBox.Show(lEx.Message, Text, MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        End Try

    End Sub

    Private Sub btnDisconnect_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles disconnectButton.Click
        ' Show hourglass.
        Dim lPrevCursor As Cursor = Cursor
        Cursor = Cursors.WaitCursor

        ' Disconnect from device.
        Disconnect()

        ' Restore cursor.
        Cursor = lPrevCursor
    End Sub

    Private Sub btnPlay_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles playButton.Click
        StartAcquisition()
    End Sub

    Private Sub btnStop_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles stopButton.Click
        StopAcquisition()
    End Sub

    Private Sub btnCommunication_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles communicationButton.Click
        ShowGenWindow(mCommBrowser, mDevice.CommunicationParameters, "Communication Control")
    End Sub

    Private Sub btnDevice_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles deviceButton.Click
        ShowGenWindow(mDeviceBrowser, mDevice.Parameters, "Device Control")
    End Sub

    Private Sub btnStream_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles streamButton.Click
        ShowGenWindow(mStreamBrowser, mStream.Parameters, "Image Stream Control")
    End Sub

    Private Sub cbxMode_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles modeComboBox.SelectedIndexChanged
        'If nothing selected, do nothing.
        If modeComboBox.SelectedIndex < 0 Then
            Return
        End If

        Dim lMode As PvGenEnum = mDevice.Parameters.GetEnum("AcquisitionMode")
        ' Take selection, propagate to parameter.
        Dim lSelected As String = CType(modeComboBox.SelectedItem, String)
        lMode.ValueString = modeComboBox.SelectedItem.ToString()
    End Sub

    Private Sub MainForm_FormClosing(ByVal sender As System.Object, ByVal e As System.Windows.Forms.FormClosingEventArgs) Handles MyBase.FormClosing
        Disconnect()
    End Sub

End Class
